flowmachine/flowmachine/core/server/query_schemas/per_subscriber_aggregate.py

```diff
@@ -0,0 +1,37 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from functools import reduce
+
+from marshmallow import fields
+from marshmallow.validate import OneOf
+
+from flowmachine.core.server.query_schemas import BaseExposedQuery
+from flowmachine.core.server.query_schemas.subscriber_subset import SubscriberSubset
+from flowmachine.core.server.query_schemas.base_schema import BaseSchema
+from flowmachine.features.subscriber.per_subscriber_aggregate import (
+    PerSubscriberAggregate,
+    agg_methods,
+)
+
+
+class PerSubscriberAggregateExposed(BaseExposedQuery):
+    def __init__(self, subscriber_queries, agg_method):
+        self.subscriber_query = reduce(lambda x, y: x.join(y), subscriber_queries)
+        self.agg_method = agg_method
+
+    def _flomachine_query_obj(self):
+        return PerSubscriberAggregate(
+            subscriber_query=self.subscriber_query,
+            agg_column="value",
+            agg_method=self.agg_method,
+        )
+
+
+class PerSubscriberAggregateSchema(BaseSchema):
+    query_kind = fields.String(validate=OneOf(["per_subscriber_aggregate"]))
+    # TODO: Can we make this a list or a single query?
+    subscriber_query = fields.List(SubscriberSubset)
```

Comment:
I'm not sure whether `fields.Nested(NumericSubscriberMetricsSchema, many=True)` enforces that all elements of the list have the same query kind (e.g. "total_active_periods"), or whether this would allow a list of heterogeneous query kinds - I guess we'd want the former, so it's worth checking that.

Reply:
Would `fields.List(fields.Nested(NumericSubscriberMetrics), validate=Length(min=1))` be better here?

Reply:
To be honest, I'm not sure what the difference is between `fields.List(fields.Nested(...))` and `fields.Nested(..., many=True)` (or whether there is a difference). Ideally, I think we'd like to enforce that every item in the list is the same type of subscriber metric (because it doesn't necessarily make sense to aggregate over a union of "total_active_periods" and "topup_amount", for example, and in some cases that might not even be a valid operation - e.g. if one of the queries has additional columns, or 'value' has a different data type). I'm not sure whether either `List` or `many=True` enforces that, though.

Reply:
Also not sure whether this choice has an implication for the construction of API scopes for token generation - I'll check.

Reply:
Looks like both approaches are functionally identical - both produce the same API spec, and both allow heterogeneous lists. I'm fairly sure that the permissions-checking will also allow a heterogeneous list, provided the token gives permissions for each of the query kinds present in the list, so we can't get around the problem that way either.

So we can either find a way to enforce homogeneity (perhaps by defining a custom Field, or by using `@validates_schema`), or just accept heterogeneous lists. I think `@validates_schema` might be an appropriate approach - I'm not sure the additional constraint will be reflected in the generated API spec, but I suppose it's better to have a non-advertised validation constraint than to accept parameters that will produce a broken query.

Reply:
In this respect, is heterogeneous inputs a problem? As long as they all have a `subscriber` and `value` column, they should produce a valid result, and may be useful (say you wanted to select the maximum value from two different sets of subscriber metrics)

Reply:
If they both have _only_ 'subscriber' and 'value' columns, and the value columns are of compatible type, then heterogeneous is fine. If either query kind has additional columns then 'UNION' will fail (e.g. we might want to allow LocationVisits here once that's exposed, and that also has location columns). Similarly if the 'value' columns have different types, that cannot be implicitly cast to each other, then the queries cannot be unioned (although given we're explicitly choosing "numeric subscriber metric" queries here, this point is unlikely to be an issue). But you're right, there might be times where a heterogeneous union is actually useful. For now, provided all the currently-allowed metrics have the same columns, and the 'value' columns all have compatible numeric types, let's allow heterogeneous lists - we can revisit the question if we decide to add LocationVisits as an option here.